### PR TITLE
Modify work order

### PR DESCRIPTION
**Pull request by milosvrgovic**

### CODE REVIEW

**Review by nemanjadimsic — CHANGES_REQUESTED** [review]
Needs changes, do the more edge case tests

File: work-order-service/pom.xml [hunk 1]
```diff
@@ -100,17 +100,24 @@
...
-    		<version>2.10.6</version>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jpa</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>com.h2database</groupId>
+			<artifactId>h2</artifactId>
+		</dependency>
+		<!-- ODATA4J attemp -->
+		<dependency>
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/pom.xml]: Deleted.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/SpentMaterialController.java [hunk 2]
```diff
@@ -32,21 +34,58 @@
 	@Autowired
 	MaterialRepository materialRepo;
 	
+	@Autowired
+	SpentMaterialService spentMaterialService;
+	
 	@PostMapping("{id}")
 	public ResponseEntity<?> addSpentMaterial(@PathVariable UUID id, @RequestBody SpentMaterialDTO spentMaterialDTO){
 		WorkOrder workOrder = workOrderRepo.getOne(id);
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/SpentMaterialController.java]: Moved to service.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/service/WorkOrderService.java [hunk 3]
```diff
@@ -74,25 +83,22 @@
 	@Autowired
 	WorkOrderMachineRepository womRepo;
 	
-	public void addWorkOrder(WorkOrderDTO workOrderDTO) {
-		try {
+	@Autowired
+	SAP4HanaProxy sap4hana;
+	
+	public SAPResponse addWorkOrder(WorkOrderDTO workOrderDTO) throws Exception {
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/service/WorkOrderService.java:84]: I separated this function, so now getting csrf token and cookie is in another function. Still, there is space to improve this little bit more in the future.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/SpentMaterialController.java [hunk 4]
```diff
@@ -32,21 +34,58 @@
...
 		Material material = materialRepo.getOne(spentMaterialDTO.getMaterial().getId());
 		spentMaterial.setMaterial(material);
 		
 		spentMaterialRepo.save(spentMaterial);
 		return new ResponseEntity<>(HttpStatus.OK);
 	}
+	
+	@PostMapping("updateSpentMaterial/{id}")
+	public ResponseEntity<?> updateSpentMaterial(@PathVariable UUID id, @RequestBody SpentMaterialDTO spentMaterialDTO){
+		SpentMaterial spentMaterial = spentMaterialService.getOne(id);
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/SpentMaterialController.java]: Moved.

File: zaga-web/src/app/components/field/field.component.html [hunk 5]
```diff
@@ -1,54 +1,172 @@
...
-              class="col-md-auto"
-              style="width: fit-content; padding-top: 12px"
-            >
-              <img
-                class="card-img-right"
-                src="../../../assets/field.png"
-                width="80px"
-                height="80px"
-              />
+<div class="d-none d-xl-block">
```
Comment (nemanjadimsic) [inline zaga-web/src/app/components/field/field.component.html:1]: Master data views should be constant. For example page /masterData/crop looks good:
image
Page /masterData/field should have the same template, instead:
image

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/service/WorkOrderService.java [hunk 6]
```diff
@@ -188,33 +183,16 @@ public SAPResponse addWorkOrder(WorkOrderDTO workOrderDTO) throws Exception {
 			}
 			WorkOrder wo = getOneW(workOrderId);
 			//System.out.println(wo.getWorkers().size()+"AAA");
+			Log.info("Work order creation successfuly finished");
```
Comment (nemanjadimsic) [inline work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/service/WorkOrderService.java:186]: We don't use com.sun.tools.sjavac.Log rather Logger log.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderController.java [hunk 7]
```diff
@@ -29,16 +39,45 @@
...
+	
+	@Autowired
+	SAP4HanaProxy sap4hana;
+	
+	@Autowired
+	SAPAuthConfiguration authConfiguration;
+	
 	@PostMapping("/createWorkOrder")
-	public ResponseEntity<?> createWorkOrder(@RequestBody WorkOrderDTO request){
+	public ResponseEntity<?> createWorkOrder(@RequestBody WorkOrderDTO request) throws Exception{
```
Comment (nemanjadimsic) [inline work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderController.java:45]: Not quietly what I had in mind. Function addWorkOrder can couse many exception. Try to introduce ResponseEntityExceptionHandler  which can help handling many types of exception, custom or default and set a exception throwing standard for whole team. This is top priority :fire:

**Review by nemanjadimsic — CHANGES_REQUESTED** [review]
Some more changes are required

File: zaga-web/src/app/components/creatework-order/creatework-order.component.html [hunk 8]
```diff
@@ -828,8 +687,21 @@ <h3 *ngIf="!new">Informacije o nalogu</h3>
       <i class="fa fa-leaf mr-2"></i>Materijali<i
         *ngIf="!materials"
         class="fa fa-arrow-down mr-2 mt-1 arrows"
+        (click)="expandMaterials()"
```
[the remark this answers is not in the document]
Comment (milosvrgovic) [reply zaga-web/src/app/components/creatework-order/creatework-order.component.html:694]: Resolved

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderController.java [hunk 9]
```diff
@@ -77,13 +116,41 @@
...
-		wrepo.save(workOrder);
-		
+	@PostMapping("/closeWorkOrder")
+	public ResponseEntity<?> closeWorkOrder(@RequestBody WorkOrderDTO request){
+		workOrderService.closeWorkOrder(request);
 		return new ResponseEntity<>(HttpStatus.OK);
 	}
 
+	@GetMapping("/test")
+	public ResponseEntity<?> test(){
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderController.java:119]: Actually, we don't need this. It's deleted.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderController.java [hunk 10]
```diff
@@ -29,16 +39,45 @@
...
 	}
 	
+	@PostMapping("/createCopy/{id}")
+	public ResponseEntity<?> copyWorkOrder(@PathVariable UUID id, @RequestBody DateDTO date){
+		WorkOrder workOrder = wrepo.getOne(id);
+		
+		LocalDate newDate = new LocalDate(Integer.parseInt(date.getYear()),
+										  Integer.parseInt(date.getMonth()),
+										  Integer.parseInt(date.getDay()));
+		System.out.println(newDate.toString());
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderController.java:68]: Yes, but I moved this part of the code into service.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderWorkerController.java [hunk 11]
```diff
@@ -34,20 +37,107 @@
...
 
 	@PostMapping("addWorker/{id}")
 	public ResponseEntity<?> addWorker(@PathVariable UUID id,@RequestBody WorkOrderWorkerDTO wowDTO){
 		WorkOrder workOrder = workOrderRepo.getOne(id);
 		WorkOrderWorker wow = new WorkOrderWorker();
-		wow.setDate(new Date()); //zakucano
-		wow.setDayNightPeriod(wowDTO.getDayNightPeriod());
-		wow.setDayWorkPeriod(wowDTO.getDayPeriod());
-		wow.setWorkPeriod(wowDTO.getDayNightPeriod() + wowDTO.getDayPeriod());
+		if(wowDTO.getNightPeriod() != null) {
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderWorkerController.java:51]: It has been moved.

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderWorkerController.java [hunk 12]
```diff
@@ -34,20 +37,107 @@
...
+		}
 		wow.setWorkOrder(workOrder);
 		wow.setUser(userRepo.getOne(wowDTO.getUser().getId()));
 
 		wowRepo.save(wow);
 		return new ResponseEntity<>(HttpStatus.OK);
 	}
+	
+	@PostMapping("updateWorkOrderWorker/{id}")
+	public ResponseEntity<?> updateWorkOrderWorker(@PathVariable UUID id, @RequestBody WorkOrderWorkerDTO wowDTO){
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/controller/WorkOrderWorkerController.java:92]: It has been moved.

File: zaga-web/src/app/components/creatework-order/creatework-order.component.html [hunk 13]
```diff
@@ -270,541 +254,417 @@ <h3 *ngIf="!new">Informacije o nalogu</h3>
...
-              <th style="text-align: right !important">Dodaj</th>
-            </thead>
-            <tbody>
-              <tr>
-                <td>
-                  <div class="form-group">
+              <!--editable table worker/machine new wo-->
+              <ng-container *ngIf="new; else elseBlockMachineState">
+                <span *ngIf="wows.length == 0"
+                  >Nema dodatih radnika i mašina, dodajte klikom na dugme </span
```
[the remark this answers is not in the document]
Comment (milosvrgovic) [reply zaga-web/src/app/components/creatework-order/creatework-order.component.html:470]: Resolved

File: zaga-web/src/app/components/creatework-order/creatework-order.component.ts [hunk 14]
```diff
@@ -268,261 +266,316 @@ export class CreateworkOrderComponent implements OnInit {
...
   updateWorkOrder() {
-    var dateStartToAdd = "";
-    var dateEndToAdd = "";
-    if (this.workOrder.start != undefined) {
-      if (this.workOrder.start.month < 10) {
-        this.workOrder.start.month = "0" + this.workOrder.start.month;
-      }
-      if (this.workOrder.start.day < 10) {
-        this.workOrder.start.day = "0" + this.workOrder.start.day;
+    if (this.workOrder.date != undefined) {
```
[the remark this answers is not in the document]
Comment (milosvrgovic) [reply zaga-web/src/app/components/creatework-order/creatework-order.component.ts:734]: Resolved

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/dto/CultureDTO.java [hunk 15]
```diff
@@ -15,7 +15,7 @@
 @NoArgsConstructor
 public class CultureDTO {
 	
-	private UUID id;
+	private UUID dbId;
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/dto/CultureDTO.java:18]: Not sure if I understood this comment.

File: zaga-web/src/app/components/field/field.component.html (hunk 5 above)
Comment (milosvrgovic) [reply zaga-web/src/app/components/field/field.component.html:1]: Resolved

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/dto/DateDTO.java [hunk 16]
```diff
@@ -16,9 +16,10 @@
 	private String dayWithHours;
 	
 	public DateDTO(WorkOrder wo) {
-		dayWithHours = wo.getStartDate().toString().split("-")[2];
-		day = dayWithHours.split(" ")[0];
-		month = wo.getStartDate().toString().split("-")[1];
-		year = wo.getStartDate().toString().split("-")[0];
+		//System.out.println(wo.getDate().toString());
```
[the remark this answers is not in the document]
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/dto/DateDTO.java]: Yes, it's working correctly.

File: zaga-web/src/app/components/field/field.component.html [hunk 17]
```diff
@@ -0,0 +1,148 @@
+<div style="padding-top: 10px"></div>
+<h3 class="mt-2 mb-2">Pregled parcela</h3>
+<div style="padding-top: 20px"></div>
+<div class="containerFields">
```
[the remark this answers is not in the document]
Comment (milosvrgovic) [reply zaga-web/src/app/components/field/field.component.html]: Resolved

File: work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/service/WorkOrderService.java (hunk 6 above)
Comment (katic-milan) [reply work-order-service/src/main/java/org/mkgroup/zaga/workorderservice/service/WorkOrderService.java]: Resolved.

File: zaga-web/src/app/components/creatework-order/creatework-order.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (nemanjadimsic) [reply zaga-web/src/app/components/creatework-order/creatework-order.component.ts:735]: When copying work order this happens:

image

* autocomplete does not work
* workers and materials are not copied (maybe for a reason)
* when adding workers, they are not visible in a list

[the remark this answers is not in the document]
Comment (milosvrgovic) [reply zaga-web/src/app/components/creatework-order/creatework-order.component.ts:761]: Resolved
